File: src/main/java/org/hydev/clock_api/controller/UserController.java
package org.hydev.clock_api.controller;

import org.hydev.clock_api.entity.User;
import org.hydev.clock_api.error.ErrorCode;
import org.hydev.clock_api.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.util.DigestUtils;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import javax.validation.constraints.Pattern;
import java.util.Date;
import java.util.List;
import java.util.function.Function;

@Validated
@RestController
@RequestMapping("/user")
public class UserController {
    private final UserRepository userRepository;

    @Autowired
    public UserController(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    /**
     * Register a user to the database.
     *
     * https://www.baeldung.com/spring-rest-http-headers
     * This method should be synchronized to avoid race condition.
     * Also, this method should not be private, or else cannot use userRepository.
     *
     * TODO: 2021/1/22 Need a better design!
     * Controller Return error code list as List<String>, or return uuid as String.
     *
     * @param username Unique username (Not empty, and should match the regex {@code User.RE_USERNAME})
     * @param password Password initial hash (Not empty)
     * @return Success or error
     */
    @PostMapping("/register")
    @SuppressWarnings("rawtypes")
    public synchronized ResponseEntity register(
            // [!] @RequestHeader(required = false) makes no need make another error handler.
            // [!] And also, ExceptionHandler of MissingRequestHeaderException cannot deal with all missing fields.
            @Pattern(regexp = User.RE_USERNAME, message = ErrorCode.USER_NAME_NOT_MATCH_REGEX)
            @RequestHeader String username,
            @Pattern(regexp = User.RE_PASSWORD, message = ErrorCode.USER_PASSWORD_NOT_MATCH_REGEX)
            @RequestHeader String password
    ) {
        // First, spring will check args. If not pass there regex, raise ConstraintViolationException.
        // Then, we check if username not exists. If username already exists, return ErrorCode with 409 conflict.
        if (userRepository.existsByUsername(username))
            // TODO: 2021/1/22 Using library instead of hand make.
            return ResponseEntity.status(HttpStatus.CONFLICT).body(List.of(ErrorCode.USER_NAME_ALREADY_EXISTS));

        User user = new User();
        user.setUsername(username);
        user.setPasswordMd5(userToSaltedMd5(username, password));
        user.setJoinDate(new Date());

        // After save and flush, uuid field will be generated automatically.
        userRepository.saveAndFlush(user);
        return ResponseEntity.ok(user.getUuid());
    }

    /**
     * Create salted hash for user's password
     * Format: "$username + $password".toLowerMd5();
     *
     * @param username Unique username used as a salt
     * @param password Password initial hash
     * @return Salted hash
     */
    private static String userToSaltedMd5(String username, String password) {
        String beforeMd5 = String.format("%s + %s", username, password);
        return DigestUtils.md5DigestAsHex(beforeMd5.getBytes()).toLowerCase();
    }

    /**
     * Check username & password.
     * - User doesn't exist -> http 404
     * - Password doesn't match -> http 401
     * - All match -> Execute operation and return the resulting String.
     *
     * @param username Unique username
     * @param password Password initial hash
     * @param operation Callback on success
     * @return Callback result or the error response
     */
    private ResponseEntity<String> checkPasswordAndDo(String username, String password,
                                                      Function<User, String> operation) {
        User user = userRepository.queryByUsername(username);
        if (user == null) return ResponseEntity.notFound().build();

        if (!user.getPasswordMd5().equals(userToSaltedMd5(username, password)))
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();

        String result = operation.apply(user);
        return ResponseEntity.ok(result);
    }

    @PostMapping("/delete")
    public synchronized ResponseEntity<String> delete(@RequestHeader String username, @RequestHeader String password) {
        return checkPasswordAndDo(username, password, user -> {
            userRepository.delete(user);
            return "";
        });
    }

    @PostMapping("/login")
    public synchronized ResponseEntity<String> login(@RequestHeader String username, @RequestHeader String password) {
        return checkPasswordAndDo(username, password, user -> userRepository.queryByUsername(username).getUuid());
    }
}

File: src/main/java/org/hydev/clock_api/controller/UserController.kt
package org.hydev.clock_api.controller

import org.hydev.clock_api.*
import org.hydev.clock_api.entity.FamilyRepo
import org.hydev.clock_api.entity.User
import org.hydev.clock_api.error.ErrorCode
import org.hydev.clock_api.repository.UserRepository
import org.springframework.data.repository.findByIdOrNull
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import javax.validation.constraints.Pattern

@RestController
@RequestMapping("/user")
open class UserController(private val userRepo: UserRepository, private val familyRepo: FamilyRepo)
{
    /**
     * Register a user to the database.
     *
     *
     * https://www.baeldung.com/spring-rest-http-headers
     * This method should be synchronized to avoid race condition.
     * Also, this method should not be private, or else cannot use userRepository.
     *
     *
     * TODO: 2021/1/22 Need a better design!
     * Controller Return error code list as List<String>, or return uuid as String.
     *
     * @param username Unique username (Not empty, and should match the regex `User.RE_USERNAME`)
     * @param password Password initial hash (Not empty)
     * @return Success or error
     */
    @PostMapping("/register")
    @Synchronized
    fun register( 
        // [!] @H(required = false) makes no need make another error handler.
        // [!] And also, ExceptionHandler of MissingHException cannot deal with all missing fields.
        @H username: @Pattern(regexp = User.RE_USERNAME, message = ErrorCode.USER_NAME_NOT_MATCH_REGEX) Str,
        @H password: @Pattern(regexp = User.RE_PASSWORD, message = ErrorCode.USER_PASSWORD_NOT_MATCH_REGEX) Str): Any
    {
        // First, spring will check args. If not pass there regex, raise ConstraintViolationException.
        // Then, we check if username not exists. If username already exists, return ErrorCode with 409 conflict.
        // TODO: 2021/1/22 Using library instead of hand make.
        if (userRepo.existsByUsername(username)) return listOf(ErrorCode.USER_NAME_ALREADY_EXISTS).http(409)
        
        val user = User().apply {
            this.username = username
            passwordMd5 = userToSaltedMd5(username, password) 
        }

        // After save and flush, uuid field will be generated automatically.
        userRepo.saveAndFlush(user)
        return user.uuid
    }

    @PostMapping("/delete")
    @Synchronized
    fun delete(@H username: Str, @H password: Str): Any
    {
        return userRepo.login(username, password) { user ->

            // Remove user from family
            if (user.family != null)
            {
                val family = familyRepo.findByIdOrNull(user.family)
                if (family != null)
                {
                    family.members = family.members.csv().apply { remove(username) }.csv()
                    familyRepo.save(family)
                }
            }

            // Remove user
            userRepo.delete(user)

            // Done
            "Success"
        }
    }

    @PostMapping("/login")
    @Synchronized
    fun login(@H username: Str, @H password: Str): Any
    {
        return userRepo.login(username, password) { it.uuid }
    }

    @PostMapping("/backup/upload")
    fun backupUpload(@H username: Str, @H password: Str, @H config: Str): Any
    {
        return userRepo.login(username, password) { user ->

            user.backup = config
            userRepo.save(user)
            "Success"
        }
    }

    @PostMapping("/backup/download")
    fun backupDownload(@H username: Str, @H password: Str): Any
    {
        return userRepo.login(username, password) { user -> user.backup }
    }
}
